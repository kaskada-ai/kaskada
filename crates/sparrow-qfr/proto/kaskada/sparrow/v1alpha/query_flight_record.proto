syntax = "proto3";
package kaskada.sparrow.v1alpha;

import "google/protobuf/wrappers.proto";

option go_package = "gitlab.com/kaskada/kaskada/wren/gen/go/kaskada/sparrow/v1alpha";
option java_multiple_files = true;
option java_package = "com.kaskada.sparrow.v1alpha1";

// Metadata added at the beginning of the flight record file.
//
// Contains metadata on the query, plan, etc. useful for interepreting
// the flight record.
message FlightRecordHeader {
  // The version of the flight record format contained here.
  uint32 version = 1;

  // The request ID this flight record is for.
  string request_id = 2;

  // Build information about the Sparrow Binary.
  BuildInfo sparrow_build_info = 3;

  // Register activities that may be reported during the query.
  //
  // This includes the set of all activities which may be produced by any query.
  // Not all activities will be reported by any specific query.
  repeated RegisterActivity activities = 4;

  // Register metrics that may be reported during the query.
  //
  // Metrics may be reported as part of an activity (see ReportActivity) or
  // at the top level during execution (see ReportMetrics).
  //
  // The set of metrics are static within the query engine, and correspond
  // to any value that may be reported during the execution of a query. Not
  // all metrics need to be reported for any given query.
  repeated RegisterMetric metrics = 5;

  message BuildInfo {
    string sparrow_version = 1;
    string github_ref = 2;
    string github_sha = 3;
    string github_workflow = 4;
  }

  message RegisterActivity {
    uint32 activity_id = 1;

    string label = 2;

    // The ID of the parent activity.
    //
    // May be `None` / `null` to indicate a root activity.
    google.protobuf.UInt32Value parent_activity_id = 3;
  }

  message RegisterMetric {
    uint32 metric_id = 1;
    string label = 2;
    MetricKind kind = 3;

    // string description = 4;
    // string units = 5;

    enum MetricKind {
      METRIC_KIND_UNSPECIFIED = 0;

      METRIC_KIND_U64_GAUGE = 10;
      METRIC_KIND_I64_GAUGE = 11;
      METRIC_KIND_F64_GAUGE = 12;

      METRIC_KIND_U64_COUNTER = 20;
      METRIC_KIND_I64_COUNTER = 21;
      METRIC_KIND_F64_COUNTER = 22;
    }
  }
}

message FlightRecord {
  oneof record {
    // Register an execution thread.
    // This is sent for each computation thread spawned for a specific query.
    RegisterThread register_thread = 1;

    /// Report an activity execution during the query.
    ReportActivity report_activity = 2;

    /// Report a top-level metric.
    ReportMetrics report_metrics = 3;
  }

  message RegisterThread {
    uint32 thread_id = 1;
    string label = 2;
  }

  message ReportActivity {
    uint32 activity_id = 1;
    uint32 thread_id = 2;

    // The time within the trace at which this event was recorded.
    uint64 wall_timestamp_us = 3;

    // Wall clock time spent executing this event.
    //
    // This should be `0` for point-wise events.
    uint64 wall_duration_us = 4;

    // CPU time spent executing this event.
    uint64 cpu_duration_us = 5;

    // Metrics to report associated with the given activity.
    repeated MetricValue metrics = 6;
  }

  message ReportMetrics {
    uint32 thread_id = 1;

    // The time within the trace at which this event was recorded.
    uint64 wall_timestamp_us = 2;

    // Metrics to report against the given thread.
    repeated MetricValue metrics = 3;
  }
}

message MetricValue {
  uint32 metric_id = 1;

  oneof value {
    uint64 u64_value = 2;
    int64 i64_value = 3;
    double f64_value = 4;
  }
}