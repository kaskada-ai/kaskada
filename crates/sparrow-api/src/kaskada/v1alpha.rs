#![allow(clippy::large_enum_variant)] // Could box the pulsar subscription, but for now just allowing this.
#![allow(clippy::derive_partial_eq_without_eq)]
use std::str::FromStr; // stop clippy erroring on generated code from tonic (proto)

use itertools::Itertools;
use url::Url;

use crate::kaskada::v1alpha::object_store_destination::ResultPaths;
use crate::kaskada::v1alpha::operation_plan::tick_operation::TickBehavior;

tonic::include_proto!("kaskada.kaskada.v1alpha");

/// Traits for the `Slice`
mod slice_traits;

/// Implementations and traits for parts of the plan
mod plan_impl;

/// Traits for [PreparedFile]
mod prepared_file_impl;

/// Traits for [Schema]
mod schema_traits;

impl ComputeTable {
    pub fn name(&self) -> &str {
        &self.config().name
    }

    pub fn config(&self) -> &TableConfig {
        self.config.as_ref().expect("tables should have a config")
    }
}

impl ExecuteResponse {
    /// Returns an owned vec of the output paths for this execution.
    //
    /// Returns `None` if not an `ObjectStoreOutput`.
    pub fn output_paths(&self) -> Option<Vec<String>> {
        match &self.destination {
            Some(output) => match &output.destination {
                Some(destination::Destination::ObjectStore(store)) => store
                    .output_paths
                    .as_ref()
                    .map(|output_paths| output_paths.paths.clone()),
                Some(_) | None => None,
            },
            None => None,
        }
    }

    /// Returns a mutable reference to the output paths.
    ///
    /// Returns `None` if not an `ObjectStoreOutput`.
    pub fn output_paths_mut(&mut self) -> Option<&mut ResultPaths> {
        match &mut self.destination {
            Some(output) => match &mut output.destination {
                Some(destination::Destination::ObjectStore(store)) => match &mut store.output_paths
                {
                    Some(output_path) => Some(output_path),
                    None => None,
                },
                Some(_) | None => None,
            },
            None => None,
        }
    }
}

impl FeatureSet {
    pub fn new(query: &str, formulas: Vec<(&str, &str)>) -> Self {
        Self {
            formulas: formulas
                .into_iter()
                .enumerate()
                .map(|(index, (name, formula))| Formula {
                    name: name.to_owned(),
                    formula: formula.to_owned(),
                    source_location: format!("FeatureSet::new formula {index}"),
                })
                .collect(),
            query: query.to_owned(),
        }
    }
}

impl TableConfig {
    /// Creates a table config with the given parameters and a `KaskadaSource`,
    /// which represents a file-backed table.
    pub fn new_with_table_source(
        name: &str,
        uuid: &uuid::Uuid,
        time_column_name: &str,
        subsort_column_name: Option<&str>,
        group_column_name: &str,
        grouping: &str,
    ) -> Self {
        Self {
            name: name.to_owned(),
            uuid: uuid.to_string(),
            time_column_name: time_column_name.to_owned(),
            subsort_column_name: subsort_column_name.map(|s| s.to_owned()),
            group_column_name: group_column_name.to_owned(),
            grouping: grouping.to_owned(),
            source: Some(Source {
                source: Some(source::Source::Kaskada(KaskadaSource {})),
            }),
        }
    }

    pub fn new2(
        name: &str,
        time_column_name: &str,
        subsort_column_name: Option<&str>,
        group_column_name: &str,
        grouping: &str,
    ) -> Self {
        Self::new_with_table_source(
            name,
            &uuid::Uuid::new_v4(),
            time_column_name,
            subsort_column_name,
            group_column_name,
            grouping,
        )
    }

    pub fn with_table_source(self) -> Self {
        Self::new_with_table_source(
            &self.name,
            &uuid::Uuid::from_str(&self.uuid).unwrap(),
            &self.time_column_name,
            self.subsort_column_name.as_deref(),
            &self.group_column_name,
            &self.grouping,
        )
    }
}

impl Formula {
    pub fn new(name: &str, formula: &str) -> Self {
        Self {
            name: name.to_owned(),
            formula: formula.to_owned(),
            source_location: name.to_owned(),
        }
    }
}

// Display PlanHash as the upper hex encoded vector.
impl std::fmt::Display for PlanHash {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            self.hash
                .iter()
                .format_with("", |elt, f| f(&format_args!("{elt:X}")))
        )
    }
}

impl SourceData {
    pub fn try_from_local(path: &std::path::Path) -> anyhow::Result<source_data::Source> {
        let path = match path.extension().and_then(|ext| ext.to_str()) {
            Some("parquet") => {
                source_data::Source::ParquetPath(format!("file://{}", path.display()))
            }
            Some("csv") => source_data::Source::CsvPath(format!("file://{}", path.display())),
            unsupported => anyhow::bail!("Unsupported extension {:?}", unsupported),
        };
        Ok(path)
    }

    pub fn try_from_url(url: &Url, extension: Option<&str>) -> anyhow::Result<source_data::Source> {
        let path = match extension {
            Some("parquet") => source_data::Source::ParquetPath(url.to_string()),
            Some("csv") => source_data::Source::CsvPath(url.to_string()),
            unsupported => anyhow::bail!("Unsupported extension {:?}", unsupported),
        };
        Ok(path)
    }
}

impl std::fmt::Display for FenlDiagnostics {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            Itertools::intersperse(
                self.fenl_diagnostics
                    .iter()
                    .map(|d| -> &str { &d.formatted }),
                ""
            )
            .flat_map(|f| f.lines())
            .map(|s| s.trim_end())
            .format("\n")
        )
    }
}

impl std::fmt::Display for TickBehavior {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TickBehavior::Minutely => write!(f, "minutely"),
            TickBehavior::Hourly => write!(f, "hourly"),
            TickBehavior::Daily => write!(f, "daily"),
            TickBehavior::Monthly => write!(f, "monthly"),
            TickBehavior::Yearly => write!(f, "yearly"),
            TickBehavior::Finished => write!(f, "final"),
            TickBehavior::Unspecified => panic!("Unspecified tick behavior"),
        }
    }
}

impl From<uuid::Uuid> for Uuid {
    fn from(uuid: uuid::Uuid) -> Self {
        let (high, low) = uuid.as_u64_pair();
        Self { high, low }
    }
}

impl From<&uuid::Uuid> for Uuid {
    fn from(uuid: &uuid::Uuid) -> Self {
        let (high, low) = uuid.as_u64_pair();
        Self { high, low }
    }
}

impl From<Uuid> for uuid::Uuid {
    fn from(uuid: Uuid) -> Self {
        Self::from_u64_pair(uuid.high, uuid.low)
    }
}

impl From<&Uuid> for uuid::Uuid {
    fn from(uuid: &Uuid) -> Self {
        Self::from_u64_pair(uuid.high, uuid.low)
    }
}
