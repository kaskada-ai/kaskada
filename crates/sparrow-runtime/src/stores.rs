mod object_meta_ext;
mod object_store_key;
pub mod object_store_url;
pub mod registry;

pub use object_meta_ext::ObjectMetaExt;
pub use object_store_url::ObjectStoreUrl;
pub use registry::ObjectStoreRegistry;
