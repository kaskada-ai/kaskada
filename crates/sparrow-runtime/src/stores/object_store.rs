/// Wrapper around an ObjectStore from the `object_store` crate.
pub struct ObjectStore(Arc<dyn object_store::ObjectStore>);
