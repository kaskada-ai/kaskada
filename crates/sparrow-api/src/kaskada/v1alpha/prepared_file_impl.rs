use chrono::NaiveDateTime;

use crate::kaskada::v1alpha::PreparedFile;

#[derive(derive_more::Display, Debug)]
pub enum Error {
    #[display(fmt = "missing field '{field}")]
    MissingField { field: &'static str },
    #[display(fmt = "invalid timestamp {timestamp:?}")]
    InvalidTimestamp {
        timestamp: prost_wkt_types::Timestamp,
    },
}

impl error_stack::Context for Error {}

impl From<&PreparedFile> for PreparedFile {
    fn from(file: &PreparedFile) -> Self {
        file.clone()
    }
}

impl PreparedFile {
    pub fn min_event_time(&self) -> error_stack::Result<NaiveDateTime, Error> {
        let Some(timestamp) = self.min_event_time.as_ref() else {
            error_stack::bail!(Error::MissingField {
                field: "min_event_time"
            })
        };
        NaiveDateTime::from_timestamp_opt(timestamp.seconds, timestamp.nanos as u32).ok_or_else(
            || {
                error_stack::report!(Error::InvalidTimestamp {
                    timestamp: timestamp.clone()
                })
            },
        )
    }

    pub fn max_event_time(&self) -> error_stack::Result<NaiveDateTime, Error> {
        let Some(timestamp) = self.max_event_time.as_ref() else {
            error_stack::bail!(Error::MissingField {
                field: "max_event_time"
            })
        };
        NaiveDateTime::from_timestamp_opt(timestamp.seconds, timestamp.nanos as u32).ok_or_else(
            || {
                error_stack::report!(Error::InvalidTimestamp {
                    timestamp: timestamp.clone()
                })
            },
        )
    }
}
