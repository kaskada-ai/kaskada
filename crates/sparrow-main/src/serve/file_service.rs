use std::sync::Arc;

use error_stack::{IntoReport, ResultExt};
use sparrow_api::kaskada::v1alpha::file_service_server::FileService;
use sparrow_api::kaskada::v1alpha::Schema;
use sparrow_api::kaskada::v1alpha::{
    GetMetadataRequest, GetMetadataResponse, KafkaConfig, MergeMetadataRequest,
    MergeMetadataResponse, PulsarConfig, SourceData, SourceMetadata,
};

use sparrow_runtime::RawMetadata;

use sparrow_runtime::stores::ObjectStoreRegistry;
use tonic::Response;

use crate::serve::error_status::IntoStatus;

#[derive(Debug)]
pub(super) struct FileServiceImpl {
    object_stores: Arc<ObjectStoreRegistry>,
}

impl FileServiceImpl {
    pub fn new(object_stores: Arc<ObjectStoreRegistry>) -> Self {
        Self { object_stores }
    }
}

#[tonic::async_trait]
impl FileService for FileServiceImpl {
    #[tracing::instrument]
    async fn get_metadata(
        &self,
        request: tonic::Request<GetMetadataRequest>,
    ) -> Result<tonic::Response<GetMetadataResponse>, tonic::Status> {
        let object_store = self.object_stores.clone();
        match tokio::spawn(get_metadata(object_store, request)).await {
            Ok(result) => result.into_status(),
            Err(panic) => {
                tracing::error!("Panic during prepare: {panic}");
                Err(tonic::Status::internal("panic during prepare"))
            }
        }
    }

    #[tracing::instrument]
    async fn merge_metadata(
        &self,
        _request: tonic::Request<MergeMetadataRequest>,
    ) -> Result<tonic::Response<MergeMetadataResponse>, tonic::Status> {
        Err(tonic::Status::unimplemented(
            "merge_metadata not implemented",
        ))
    }
}

async fn get_metadata(
    object_store_registry: Arc<ObjectStoreRegistry>,
    request: tonic::Request<GetMetadataRequest>,
) -> anyhow::Result<tonic::Response<GetMetadataResponse>> {
    let request = request.into_inner();
    let source = request.source.unwrap();
    let file_metadata = match source {
        sparrow_api::kaskada::v1alpha::get_metadata_request::Source::SourceData(source_data) => {
            get_source_metadata(&object_store_registry, &source_data)
                .await
                .or_else(|e| anyhow::bail!("failed getting source metadata: {}", e))
        }
        sparrow_api::kaskada::v1alpha::get_metadata_request::Source::PulsarConfig(
            pulsar_config,
        ) => get_pulsar_metadata(&pulsar_config)
            .await
            .or_else(|e| anyhow::bail!("failed getting source metadata: {}", e)),
        sparrow_api::kaskada::v1alpha::get_metadata_request::Source::KafkaConfig(kafka_config) => {
            get_kafka_metadata(&kafka_config)
                .await
                .or_else(|e| anyhow::bail!("failed getting source metadata: {}", e))
        }
    }?;

    Ok(Response::new(GetMetadataResponse {
        source_metadata: Some(file_metadata),
    }))
}

#[derive(derive_more::Display, Debug)]
pub enum Error {
    #[display(fmt = "unable to get source path from request")]
    SourcePath,
    #[display(fmt = "schema error: '{_0}'")]
    Schema(String),
}
impl error_stack::Context for Error {}

pub(crate) async fn get_source_metadata(
    object_store_registry: &ObjectStoreRegistry,
    source: &SourceData,
) -> error_stack::Result<SourceMetadata, Error> {
    let source = source.source.as_ref().ok_or(Error::SourcePath)?;
    let metadata = RawMetadata::try_from(source, object_store_registry)
        .await
        .attach_printable_lazy(|| format!("Source: {:?}", source))
        .change_context(Error::Schema(format!(
            "unable to read schema from: {:?}",
            source
        )))?;
    get_schema_from_metadata(metadata)
}

pub(crate) async fn get_pulsar_metadata(
    pc: &PulsarConfig,
) -> error_stack::Result<SourceMetadata, Error> {
    let metadata = RawMetadata::try_from_pulsar_subscription(pc)
        .await
        .attach_printable_lazy(|| format!("Pulsar Source: {:?}", pc))
        .change_context(Error::Schema(format!(
            "unable to read schema from: {:?}",
            pc
        )))?;
    get_schema_from_metadata(metadata)
}

pub(crate) async fn get_kafka_metadata(
    kc: &KafkaConfig,
) -> error_stack::Result<SourceMetadata, Error> {
    let metadata = RawMetadata::try_from_kafka_subscription(kc)
        .await
        .attach_printable_lazy(|| format!("Kafka Source: {:?}", kc))
        .change_context(Error::Schema(format!(
            "unable to read schema from: {:?}",
            kc
        )))?;
    get_schema_from_metadata(metadata)
}

fn get_schema_from_metadata(metadata: RawMetadata) -> error_stack::Result<SourceMetadata, Error> {
    let schema = Schema::try_from(metadata.table_schema.as_ref())
        .into_report()
        .change_context(Error::Schema(format!(
            "Unable to encode schema {:?}",
            metadata.table_schema
        )))
        .attach_printable_lazy(|| {
            format!(
                "Raw Schema: {:?} Table Schema: {:?}",
                metadata.raw_schema, metadata.table_schema
            )
        })?;
    Ok(SourceMetadata {
        schema: Some(schema),
    })
}

#[cfg(test)]
mod tests {
    use sparrow_api::kaskada::v1alpha::source_data;

    use super::*;

    #[tokio::test]
    async fn test_timestamp_with_timezone_gets_metadata() {
        // Regression test for handling files with a timestamp with a configured
        // timezone.
        let path = sparrow_testing::testdata_path("eventdata/sample_event_data.parquet");
        let object_store_registry = Arc::new(ObjectStoreRegistry::new());
        let file_service = FileServiceImpl::new(object_store_registry);
        let result = file_service
            .get_metadata(tonic::Request::new(GetMetadataRequest {
                source: Some(
                    sparrow_api::kaskada::v1alpha::get_metadata_request::Source::SourceData(
                        SourceData {
                            source: Some(source_data::Source::ParquetPath(format!(
                                "file:///{}",
                                path.display()
                            ))),
                        },
                    ),
                ),
            }))
            .await
            .unwrap()
            .into_inner();

        insta::assert_yaml_snapshot!(result);
    }

    #[tokio::test]
    async fn test_get_metadata_csv_path() {
        let path =
            sparrow_testing::testdata_path("eventdata/2c889258-d676-4922-9a92-d7e9c60c1dde.csv");

        let object_store_registry = Arc::new(ObjectStoreRegistry::new());
        let file_service = FileServiceImpl::new(object_store_registry);
        let result = file_service
            .get_metadata(tonic::Request::new(GetMetadataRequest {
                source: Some(
                    sparrow_api::kaskada::v1alpha::get_metadata_request::Source::SourceData(
                        SourceData {
                            source: Some(source_data::Source::CsvPath(format!(
                                "file:///{}",
                                path.display()
                            ))),
                        },
                    ),
                ),
            }))
            .await
            .unwrap()
            .into_inner();
        insta::assert_yaml_snapshot!(result);
    }

    #[tokio::test]
    async fn test_get_metadata_csv_data() {
        let csv_data = "id,name,value,value2\n1,awkward,taco,123\n2,taco,awkward,456\n";
        let object_store_registry = Arc::new(ObjectStoreRegistry::new());
        let file_service = FileServiceImpl::new(object_store_registry);
        let result = file_service
            .get_metadata(tonic::Request::new(GetMetadataRequest {
                source: Some(
                    sparrow_api::kaskada::v1alpha::get_metadata_request::Source::SourceData(
                        SourceData {
                            source: Some(source_data::Source::CsvData(csv_data.to_owned())),
                        },
                    ),
                ),
            }))
            .await
            .unwrap()
            .into_inner();
        insta::assert_yaml_snapshot!(result);
    }
}
