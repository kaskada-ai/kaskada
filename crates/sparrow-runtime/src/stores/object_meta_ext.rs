use object_store::ObjectMeta;

use crate::DETERMINISTIC_RUNTIME_HASHER;

pub trait ObjectMetaExt {
    /// Return a deterministic hash of the etag.
    ///
    /// For object stores that don't provide an etag, this will return a hash of
    /// the last modified time.
    fn etag_hash(&self) -> u64;
}

impl ObjectMetaExt for ObjectMeta {
    fn etag_hash(&self) -> u64 {
        match &self.e_tag {
            None => {
                // The underlying store doesn't provide an etag.
                //
                // We generate the hash of the etag by hashing the last_modified
                // time. Technically, we could also generate the etag, but this
                // skips a step.
                //
                // In systems, the entity tag may be generated from the last
                // modified time:
                //
                // > Entity tag that uniquely represents the requested resource.
                // > ... The method by which ETag values are generated is not
                // > specified. Typically, the ETag value is a hash of the
                // > content, a hash of the last modification timestamp, or just
                // > a revision number.
                //
                // -- https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag
                DETERMINISTIC_RUNTIME_HASHER.hash_one(self.last_modified)
            }
            Some(e_tag) => DETERMINISTIC_RUNTIME_HASHER.hash_one(e_tag),
        }
    }
}
