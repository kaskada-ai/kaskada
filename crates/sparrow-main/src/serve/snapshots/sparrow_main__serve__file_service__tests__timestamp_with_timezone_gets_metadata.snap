---
source: crates/sparrow-main/src/serve/file_service.rs
expression: result
---
source_metadata:
  schema:
    fields:
      - name: anonymousId
        data_type:
          kind:
            Primitive: 14
        nullable: true
      - name: type
        data_type:
          kind:
            Primitive: 14
        nullable: true
      - name: timestamp
        data_type:
          kind:
            Primitive: 14
        nullable: true
      - name: subsort_id
        data_type:
          kind:
            Primitive: 6
        nullable: true
      - name: context_locale
        data_type:
          kind:
            Primitive: 14
        nullable: true
      - name: projectId
        data_type:
          kind:
            Primitive: 14
        nullable: true
      - name: event
        data_type:
          kind:
            Primitive: 14
        nullable: true

